btree/btree.c

```diff
@@ -3711,6 +3727,63 @@ int64_t btree_open_tick(struct m0_sm_op *smop)
 	}
 }
 
+/**
+ * btree_close_tree_tick function is used to traverse through different states to
+ * facilitate the working of m0_btree_close().
+ *
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
+ */
+int64_t btree_close_tree_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop     = M0_AMB(bop, smop, bo_op);
+	struct m0_btree       *tree    = bop->bo_arbor;
+	struct nd             *nd_head = ndlist_tlist_head(&tree->t_desc->
+							   t_active_nds);
+	struct nd             *nd_curr;
+
+	switch (bop->bo_op.o_sm.sm_state) {
+		case P_INIT:
+		if (tree->t_desc->t_ref > 1)
+			tree_put(tree->t_desc);
+		else if (tree->t_desc->t_ref == 1) {
+			nd_curr = nd_head;
+			if (nd_curr->n_starttime == 0)
+				nd_curr->n_starttime = m0_time_now();
+			if (ndlist_tlist_length(&tree->t_desc->t_active_nds) >
+			    1) {
+				/** This code is meant for debugging.
+				 *  In future, this case needs to be
+				 *  handled in a better way.
+				 */
+				if (m0_time_seconds(m0_time_now() -
+					nd_curr->n_starttime) > 5) {
+					nd_curr->n_starttime = 0;
+					return M0_ERR(-ETIMEDOUT);
+				}
+				return P_INIT;
+			} else if (nd_curr == tree->t_desc->t_root) {
+				if (nd_curr->n_ref > 1) {
+					if (m0_time_seconds(m0_time_now() -
```

Comment:
There may be a scenario when the 2 threads have called tree_close() simultaenously. Main thread is a bit slow and is waiting for the secondary thread to complete its tree_close call.
Until the secondary thread completes, we need to wait. Hence, the check for node's reference

Reply:
For that you can have ndlist_tlist_length(&tree->t_desc->t_active_nds) >1  || (nd_curr->n_ref > 1) at line no. 3753  and remove if present at 3766 instead of adding separate part just to check (nd_curr->n_ref > 1) and doing the same which are doing in previous block, 
Please check it once.

Reply:
`ndlist_tlist_length(&tree->t_desc->t_active_nds) >1 || (nd_curr->n_ref > 1)`
That wont go logically but I understand what you are saying.
I am adding a state where timeout will be checked so duplication of code is avoided.
If possible I will remove that check itself as this check ideally should be part of node_put()